### DISCUSSION TITLE
Persistent client connection

### DISCUSSION THREAD

**Discussion by FrankBJensen** [eModbus/eModbus#37] [General]
Hi there.

I have been running
https://github.com/andresarmento/modbus-arduino/blob/master/libraries/ModbusIP/ModbusIP.h
with a W5500 cabled connection, connected to a WAGO PLC via TCP, using ESP32. 
I have used the TCP_KEEP_ALIVE option, to make a persistent connection, and it seems to run stable.
Using the TCP_STOP makes the client disconnect, and that results in an modbus error state in the PLC, and it must recover some seconds, before a new connection is possible.

I want to emmigrate this to wireless WiFi.

But no matter what Modbus library I seem to find, there is no option to keep the client connection alive. It always disconnects after sendt packages. I am not sure, if its a core part of WiFi.h, I can not seem to find any workaround for this.

I do not know, what is normal practice, to make a new client for each request, or re-use the same one. I want to update my modbus variables at least every 200mS, and 100mS would we even better.

Is there a workaround for this issue, like the TCP_STOP to TCP_KEEP_ALIVE. Or is it just my PLC that is particular sensitive about this - eg. the ModScan test software so not seem to care, if connection is lost or not.

Hope for some input - or maybe even a fix so that this works for me 👍 :)

Frank

**Comment by bertmelis** [comment]
The clients in this lib don't close the connection by themselves in a typical application. It does close after a certain (configurable) "idle" time. This is for both the sync and async flavour.

Are you sure your server doesn't close the connection?

> The modbus specification says you should not open and close the connection for each request.

**Reply by Miq1** [reply]
Moved this to Discussions as it is no issue with the eModbus library.

**Comment by FrankBJensen** [comment]
To be honest, I am not quite sure, what is going on......

My best lead is, that running with cable (and W5500), I need to disable the client.stop();
If I dont do that, I have Modbus errors in my PLC. The same kind of errors, that I see using WiFi.h, in stead of Ethernet.h. And it does not matter, if I update pr. 200mS or 1000mS, of if I select the timeout in the PLC to be 10mS or 1000mS.

And why should my PLC behave differently using a cabled connection with Ethernet.h, than using WiFi with WiFi.h. The only difference is the connection, the code is the same. Except that with WiFI.h it seems to loose the connection, no matter if I have client.stop() or not.

I am quite puzzled with this, I have used several days, but I am stuck.....
You think, that its my PLC dropping the connection? (it is the slave, ESP32 is master). The PLC polls data every 200mS.

Frank.

**Reply by bertmelis** [reply]
I cannot comment on a different library. I could be a lot of things.

To the best of my knowledge, this lib honors the modbus spec. So I can only advice on using this :smile:

**Comment by FrankBJensen** [comment]
Understandable :)

Maybe I should adress this problem to WAGO, the PLC manufacturer. Since the ModScan software runs perfectly in all situations, maybe the problem is with the PLC somehow not complying to the standard.

Thank you for your time.
